### PR TITLE
Implement sample app generation for GAPIC Java

### PR DESCRIPTION
**Pull request by shinfan**
Note: This PR requires code changes to the Artman and googleapis

This PR implements the sample app generator for Java. Users can enable the sample app artifacts by including the `sample_app` in the `--enabled_artifacts` flag

**The major changes are:**
- Implement `JavaGapicSampleAppTransformer` and its related templates/transformers
- Reuse smoke test insfrastructure since the sample app is similar to the smoke test
- Add `StaticFileRunner` which allows GAPIC providers to handle static files (such as gradle wrapper)
- Make `JavaPackageCopier` implements `StaticFileRunner` so the class can be reused in the GAPIC context.
- Add `DependencyType` which indicates if the dependencies should be declared as local or release
- Baseline tests

**What's inside a sample application output:**
- Root
  - gapic src
  - build.gradle
  - settings.gradle
  - gradle wrappers
  - proto-package
  - gRPC-package

Users / Artman can run: `./gradlew jar` which will compile the generated artifacts into a fat executable jar file that is deliverable to the DPEs.

cc/ @lukesneeringer @bjwatson

### CODE REVIEW

File: src/test/java/com/google/api/codegen/testdata/java_sample_library.baseline [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  mavenLocal()
+}
+
+dependencies {
+  compile project(":grpc-google-cloud-library-v1"),
+    'com.google.api:gax:1.0.0',
+    'com.google.api:gax-grpc:0.18.0',
+    'commons-cli:commons-cli:1.4',
+    'commons-lang:commons-lang:2.6',
+    'io.netty:netty-tcnative-boringssl-static:1.1.33.Fork26'
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_sample_library.baseline]: I feel like this needs a comment telling the user they have to keep it in sync with the version that grpc needs.

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSampleAppTransformer.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import com.google.common.collect.Lists;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Responsible for producing sample application related views for Java GAPIC clients */
+public class JavaGapicSampleAppTransformer implements ModelToViewTransformer {
+
+  // Reuse the smoke test template since the sample application is almost the same as the smoke test class.
+  // (with a few exceptions such as class name)
+  private static String SMOKE_TEST_TEMPLATE_FILE = "java/smoke_test.snip";
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSampleAppTransformer.java]: Maybe we should rename the snippet file to something more generic so people don't expect it to only be used for smoke tests.

File: src/main/java/com/google/api/codegen/gapic/StaticGapicProvider.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  public StaticGapicProvider(StaticFileRunner staticFileRunner) {
+    this.staticFileRunner = staticFileRunner;
+  }
+
+  @Override
+  public Map<String, Doc> generate() {
+    try {
+      staticFileRunner.run();
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/gapic/StaticGapicProvider.java]: I'm concerned about not failing the generation process here.

File: src/test/java/com/google/api/codegen/testdata/java_sample_library.baseline (hunk 1 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_sample_library.baseline]: Comment added. Done.

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSampleAppTransformer.java (hunk 2 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSampleAppTransformer.java]: Renamed to `smoke_sample` so it looks okay for both cases. Done.

File: src/main/java/com/google/api/codegen/gapic/StaticGapicProvider.java (hunk 3 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/gapic/StaticGapicProvider.java]: Throw a runtime exception instead. Done.

**Comment by shinfan** [comment]
@garrettjonesgoogle  PTAL

File: src/main/java/com/google/api/codegen/gapic/StaticGapicProvider.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  @Override
+  public List<String> getSnippetFileNames() {
+    // Static provider does not support snippet files
+    return new ArrayList<>();
+  }
+
+  @Override
+  public Map<String, Doc> generate(String snippetFileName) {
+    // Static provider does not support snippet files
+    return new HashMap<>();
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/gapic/StaticGapicProvider.java]: I suggest `Collections.emptyMap` and `Collections.emptyList`

File: src/main/java/com/google/api/codegen/grpcmetadatagen/java/JavaPackageMetadataTransformer.java [hunk 5]
```diff
@@ -32,28 +32,35 @@
...
             config.packageName(TargetLanguage.JAVA), config.generationLayer());
-    ArrayList<PackageMetadataView> views = new ArrayList<>();
+
+    ArrayList<PackageMetadataView.Builder> viewBuilders = new ArrayList<>();
     for (String template : getSnippetsOutput().keySet()) {
-      PackageMetadataView view =
+      PackageMetadataView.Builder viewBuilder =
           metadataTransformer
               .generateMetadataView(
                   config, model, template, getSnippetsOutput().get(template), TargetLanguage.JAVA)
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/grpcmetadatagen/java/JavaPackageMetadataTransformer.java]: Use `getSnippetsOutput().entrySet()` instead?

File: src/main/java/com/google/api/codegen/gapic/StaticGapicProvider.java (hunk 4 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/gapic/StaticGapicProvider.java]: Done

File: src/main/java/com/google/api/codegen/grpcmetadatagen/java/JavaPackageMetadataTransformer.java (hunk 5 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/grpcmetadatagen/java/JavaPackageMetadataTransformer.java]: Done

File: src/main/resources/com/google/api/codegen/java/sample.gradle.snip [hunk 6]
```diff
@@ -24,6 +24,7 @@ dependencies {
     'com.google.api:gax-grpc:{@metadata.gaxGrpcVersionBound.lower}',
     'commons-cli:commons-cli:1.4',
     'commons-lang:commons-lang:2.6',
+    // This dependency needs to be update-to-date with the gRPC
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/sample.gradle.snip]: "the gRPC" doesn't make sense in isolation. I think you want "the version that gRPC expects"

Comment (shinfan) [inline src/main/resources/com/google/api/codegen/java/sample.gradle.snip]: Done

**Comment by shinfan** [comment]
I will merge this PR after the Artman PR is LGTM'd: https://github.com/googleapis/artman/pull/220

**Pull request merged**